zboxcore/sdk/downloadworker.go

```diff
@@ -46,338 +45,526 @@ type DownloadRequest struct {
+
+	var remainingMask zboxutil.Uint128
+	activeBlobbers := mask.CountOnes()
+	if activeBlobbers < requiredDownloads {
+		return zboxutil.NewUint128(0), 0, errors.New("insufficient_blobbers",
+			fmt.Sprintf("Required downloads %d, remaining active blobber %d",
+				req.consensusThresh, activeBlobbers))
+	}
+	rspCh := make(chan *downloadBlock, requiredDownloads)
+
+	var pos uint64
+	var c int
+
+	for i := req.downloadMask; !i.Equals64(0); i = i.And(zboxutil.NewUint128(1).Lsh(pos).Not()) {
+		pos = uint64(i.TrailingZeros())
+		blockDownloadReq := &BlockDownloadRequest{
+			allocationID:       req.allocationID,
+			allocationTx:       req.allocationTx,
+			allocOwnerID:       req.allocOwnerID,
+			authTicket:         req.authTicket,
+			blobber:            req.blobbers[pos],
+			blobberIdx:         int(pos),
+			chunkSize:          req.chunkSize,
+			blockNum:           startBlock,
+			contentMode:        req.contentMode,
+			result:             rspCh,
+			ctx:                req.ctx,
+			remotefilepath:     req.remotefilepath,
+			remotefilepathhash: req.remotefilepathhash,
+			numBlocks:          totalBlock,
+			encryptedKey:       req.encryptedKey,
+		}
 
-			if !blockSuccess {
-				continue
+		go AddBlockDownloadReq(blockDownloadReq)
+		c++
+		if c == requiredDownloads {
+			remainingMask = i
+			break
+		}
+
+	}
+
+	var failed int
+	failedMu := &sync.Mutex{}
+	wg := &sync.WaitGroup{}
+	for i := 0; i < requiredDownloads; i++ {
+		result := <-rspCh
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			var err error
+			defer func() {
+				if err != nil {
+					failedMu.Lock()
+					failed++
+					failedMu.Unlock()
+					req.removeFromMask(uint64(result.idx))
+					logger.Logger.Error(err)
+				}
+			}()
+			if !result.Success {
+				err = fmt.Errorf("Unsuccessful download. Error: %v", result.err)
+				return
 			}
+			err = req.fillShards(shards, result)
+			return
+		}()
+	}
+
+	wg.Wait()
+	return remainingMask, failed, nil
+}
+
+//decodeEC will reconstruct shards and verify it
+func (req *DownloadRequest) decodeEC(shards [][]byte) (data []byte, isValid bool, err error) {
+	err = req.ecEncoder.Reconstruct(shards)
+	if err != nil {
+		return
+	}
+
+	isValid, err = req.ecEncoder.Verify(shards)
+	if err != nil || !isValid {
+		return
+	}
+
+	c := len(shards[0])
+	data = make([]byte, req.datashards*c)
+	for i := 0; i < req.datashards; i++ {
+		index := i * c
+		copy(data[index:index+c], shards[i])
+	}
+	return data, true, nil
+}
 
-			success++
-			if success >= req.datashards {
-				decodeNumBlocks = downloadChunks
-				break
+// fillShards will fill `shards` with data from blobbers that belongs to specific
+// blockNumber and blobber's position index in an allocation
+func (req *DownloadRequest) fillShards(shards [][][]byte, result *downloadBlock) (err error) {
+	for i := 0; i < len(result.BlockChunks); i++ {
+		var data []byte
+		if req.encryptedKey != "" {
+			data, err = req.getDecryptedData(result, i)
```

Comment:
`req.getDecryptedData` is not thread safe, the `req.authTicket` is read without lock protection. Also, the `req.encScheme` seems not thread safe as well.